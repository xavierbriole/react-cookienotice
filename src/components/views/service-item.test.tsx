import { act, render } from '@testing-library/react'
import { vi } from 'vitest'

import ServiceItem from './service-item'

describe('ServiceItem', () => {
  beforeEach(() => {
    vi.resetAllMocks()
  })

  describe('should render', () => {
    it('with always active', () => {
      const { asFragment } = render(
        <ServiceItem
          name='serviceName'
          description='serviceDescription'
          code='serviceCode'
          alwaysActive={true}
          alwaysActiveLabel='alwaysActiveLabel'
          isChecked={false}
          onCheckboxChange={() => {}}
        />,
      )

      expect(asFragment()).toMatchSnapshot()
    })

    it('without always active', () => {
      const { asFragment } = render(
        <ServiceItem
          name='serviceName'
          description='serviceDescription'
          code='serviceCode'
          alwaysActive={false}
          alwaysActiveLabel='alwaysActiveLabel'
          isChecked={false}
          onCheckboxChange={() => {}}
        />,
      )

      expect(asFragment()).toMatchSnapshot()
    })

    it('with checked', () => {
      const { asFragment } = render(
        <ServiceItem
          name='serviceName'
          description='serviceDescription'
          code='serviceCode'
          alwaysActive={false}
          alwaysActiveLabel='alwaysActiveLabel'
          isChecked={true}
          onCheckboxChange={() => {}}
        />,
      )

      expect(asFragment()).toMatchSnapshot()
    })
  })

  it('should toggle open class', () => {
    const { getByText } = render(
      <ServiceItem
        name='serviceName'
        description='serviceDescription'
        code='serviceCode'
        alwaysActive={false}
        alwaysActiveLabel='alwaysActiveLabel'
        isChecked={false}
        onCheckboxChange={() => {}}
      />,
    )

    const button = getByText('serviceName')

    expect(button.className).not.toContain('open')

    act(() => {
      button.click()
    })

    expect(button.className).toContain('open')

    act(() => {
      button.click()
    })

    expect(button.className).not.toContain('open')
  })

  it('should add maxHeight style', () => {
    const { getByText } = render(
      <ServiceItem
        name='serviceName'
        description='serviceDescription'
        code='serviceCode'
        alwaysActive={false}
        alwaysActiveLabel='alwaysActiveLabel'
        isChecked={false}
        onCheckboxChange={() => {}}
      />,
    )

    const button = getByText('serviceName')
    const content = button.nextElementSibling

    if (!(content instanceof HTMLElement)) {
      throw new Error('content is not HTMLElement')
    }

    expect(content.style.maxHeight).toBe('0')

    act(() => {
      button.click()
    })

    expect(content.style.maxHeight).toBe(`${content.scrollHeight}px`)

    act(() => {
      button.click()
    })

    expect(content.style.maxHeight).toBe('0')
  })

  it('should handle checkbox change', () => {
    const onCheckboxChange = vi.fn()

    const { getByLabelText } = render(
      <ServiceItem
        name='serviceName'
        description='serviceDescription'
        code='serviceCode'
        alwaysActive={false}
        alwaysActiveLabel='alwaysActiveLabel'
        isChecked={false}
        onCheckboxChange={onCheckboxChange}
      />,
    )

    act(() => {
      getByLabelText('serviceCode').click()
    })

    expect(onCheckboxChange).toHaveBeenCalledTimes(1)
    expect(onCheckboxChange).toHaveBeenCalledWith(expect.any(Object))
  })
})
