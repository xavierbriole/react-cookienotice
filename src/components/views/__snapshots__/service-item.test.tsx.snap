// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ServiceItem > should render > with always active 1`] = `
<DocumentFragment>
  <div
    class="react-cookienotice-service"
  >
    <div
      class="react-cookienotice-service-checkbox"
    >
      <input
        checked=""
        disabled=""
        id="serviceCode"
        name="serviceCode"
        type="checkbox"
      />
      <label
        for="serviceCode"
      >
        serviceCode
      </label>
    </div>
    <div
      class="react-cookienotice-service-info"
    >
      <button
        class="react-cookienotice-service-info-title"
      >
        serviceName
        <div
          class="react-cookienotice-service-info-title-always-active"
        >
          <span
            class="react-cookienotice-label"
          >
            alwaysActiveLabel
          </span>
        </div>
      </button>
      <div
        class="react-cookienotice-service-info-content"
        style="max-height: 0;"
      >
        <span
          class="react-cookienotice-label"
        >
          serviceDescription
        </span>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`ServiceItem > should render > with checked 1`] = `
<DocumentFragment>
  <div
    class="react-cookienotice-service"
  >
    <div
      class="react-cookienotice-service-checkbox"
    >
      <input
        checked=""
        id="serviceCode"
        name="serviceCode"
        type="checkbox"
      />
      <label
        for="serviceCode"
      >
        serviceCode
      </label>
    </div>
    <div
      class="react-cookienotice-service-info"
    >
      <button
        class="react-cookienotice-service-info-title"
      >
        serviceName
      </button>
      <div
        class="react-cookienotice-service-info-content"
        style="max-height: 0;"
      >
        <span
          class="react-cookienotice-label"
        >
          serviceDescription
        </span>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`ServiceItem > should render > without always active 1`] = `
<DocumentFragment>
  <div
    class="react-cookienotice-service"
  >
    <div
      class="react-cookienotice-service-checkbox"
    >
      <input
        id="serviceCode"
        name="serviceCode"
        type="checkbox"
      />
      <label
        for="serviceCode"
      >
        serviceCode
      </label>
    </div>
    <div
      class="react-cookienotice-service-info"
    >
      <button
        class="react-cookienotice-service-info-title"
      >
        serviceName
      </button>
      <div
        class="react-cookienotice-service-info-content"
        style="max-height: 0;"
      >
        <span
          class="react-cookienotice-label"
        >
          serviceDescription
        </span>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
